plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -0,0 +1,129 @@
+
+    private final S3BlobStore blobStore;
+    private final String blobKey;
+    private final int maxAttempts;
+
+    private InputStream currentStream;
+    private long currentOffset;
+
+    S3RetryingInputStream(S3BlobStore blobStore, String blobKey) throws IOException {
+        this.blobStore = blobStore;
+        this.blobKey = blobKey;
+        this.maxAttempts = blobStore.getMaxRetries() + 1;
+        currentStream = openStream();
+    }
+
+    private InputStream openStream() throws IOException {
+        try (AmazonS3Reference clientReference = blobStore.clientReference()) {
+            final GetObjectRequest getObjectRequest = new GetObjectRequest(blobStore.bucket(), blobKey);
+            if (currentOffset > 0) {
+                getObjectRequest.setRange(currentOffset);
+            }
+            final S3Object s3Object = SocketAccess.doPrivileged(() -> clientReference.client().getObject(getObjectRequest));
+            return s3Object.getObjectContent();
+        } catch (final AmazonClientException e) {
+            if (e instanceof AmazonS3Exception) {
+                if (404 == ((AmazonS3Exception) e).getStatusCode()) {
+                    throw new NoSuchFileException("Blob object [" + blobKey + "] not found: " + e.getMessage());
+                }
+            }
+            throw e;
+        }
+    }
+
+    @Override
+    public int read() throws IOException {
+        int attempt = 0;
+        while (true) {
+            attempt += 1;
+            try {
+                final int result = currentStream.read();
+                currentOffset += 1;
+                return result;
+            } catch (IOException e) {
+                if (attempt >= maxAttempts) {
+                    throw e;
+                }
+                logger.debug(new ParameterizedMessage("failed reading [{}/{}] at offset [{}], attempt [{}] of [{}], retrying",
+                    blobStore.bucket(), blobKey, currentOffset, attempt, maxAttempts), e);
+                currentStream.close();
+                currentStream = openStream();
+            }
+        }
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        int attempt = 0;
+        while (true) {
+            attempt += 1;
+            try {
+                final int bytesRead = currentStream.read(b, off, len);
+                if (bytesRead == -1) {
+                    return -1;
+                }
+                currentOffset += bytesRead;
+                return bytesRead;
+            } catch (IOException e) {
+                if (attempt >= maxAttempts) {
+                    throw e;
+                }
+                logger.debug(new ParameterizedMessage("failed reading [{}/{}] at offset [{}], attempt [{}] of [{}], retrying",
+                    blobStore.bucket(), blobKey, currentOffset, attempt, maxAttempts), e);
+                currentStream.close();
```

Comment:
Should we use `IOUtils#closeWhileHandlingException` or so and suppress whatever exceptions this might throw?
I don't see what exceptions we could run into here right now, but it would be annoying if some unexpected future change destroyed out ability to retry here?

Reply:
Not sure myself - I was wondering if maybe it's right to treat an exception here as fatal. Hard to say without actually having a specific case where an exception is thrown here. I did as you suggest in ebd4e08e22cebe9547ce0ea99b802f177ac74829.